### ISSUE TITLE
Websocket chokes when sending large number of messages

### ISSUE THREAD

**Issue by mvkra** [SuaveIO/suave#519]
Upon connection to a websocket if the websocket server sends the client 1000's of messages at some random point just stops sending. From a usage perspective all I can see if that the ws.send async inside the socket ce never returns at all and blocks the rest of the application. Tried so far two different clients and am experiencing the same issue; trying another websocket framework allows it to work as per normal.
- Client connects to a websocket endpoint
- The function provided to "handshake" webpart in the "socket" computation expressions simply continues to spam the client with small messages (e.g 1, 2, 3, 4) or a repeated message
- Eventually the client will stop receiving messages and the server will be stuck.

Adding the below test case to WebSocket.fs in Suave.Tests reproduces the issue:

```
testCase "send 10000 messages to client" <| fun _ ->
  let ctx = runWithConfig webPart

  withContext (fun _ ->
    let mre = new ManualResetEvent(false)
    let amountOfMessages = 10000
    let message = "BinaryRequest16bit"
    let echo : byte array ref = ref [||]
    let count = ref 0

    use clientWebSocket = new WebSocketSharp.WebSocket(sprintf "ws://%s:%i/websocket" ip port)

    clientWebSocket.OnMessage.Add(fun e ->
      echo:= e.RawData
      count := !count + 1
      printfn "Count is %i" !count
      if !count = amountOfMessages then mre.Set() |> ignore
    )
    clientWebSocket.Connect()
    for i = 1 to amountOfMessages do
        clientWebSocket.Send(message)

    mre.WaitOne() |> ignore
    clientWebSocket.Close()
    Assert.Equal("should be equal", testByteArray (int PayloadSize.Bit16) echo.Value , true)
    Assert.Equal("received message count on websocket", amountOfMessages, !count)
    ) ctx
```

Also created a repo to reproduce the issue - simply run the console program: https://github.com/mvkra/WebsocketFrameworkTester

Bug happens on both Windows 7 Net 4.6 and Mono 4.6.1.5 so it isn't a runtime issue. At rare times it passes and different test runs get through either more or less messages before it stalls. If you can't reproduce comment out the (printfn "Count is %i" !count) and/or up the message count required.

If any other info is required happy to assist.

**Comment by mvkra** [comment]
As an aside created a transport using TcpListener instead to see if it was all the code in the SocketAsyncEventArgs transport and still got the issue. I had to up the count to around 30000 however.

For TcpListener transport see: https://gist.github.com/mvkra/ef0cde995e7d68f237f5a40d71e7cf6c

**Comment by mvkra** [comment]
Ok I think I've figured out the issue. Both the write loop and the read loop are affected by this issue which makes it even harder to track. Worse it isn't caused by any SocketError or the like so no error reporting catches it.

In a nutshell the state of either the read or write queue and whether the thread should continue or is finished is not set atomically in the readLoop or writeLoop in Websocket.fs. This is bad because their is a chance of the following happening:

- Thread 1 queues up a job (ws.send)
- The write queue is added to and the write loop is started.
- Checks the queue and sends the data. Then recursively calls writeLoop
- Checks the queue again - it is zero so it switches to setting the reading/writing variable to false (this is not done in a lock). However the setting of the variable isn't done yet.
- In between checking the queue and setting the mutable variable another ws.send is called and locks the queue. It sees their is a thread waiting and therefore doesn't create another - simply queues the send on the write queue.
- The writeLoop then ends with the item still on the queue.
- Because you are using Async.FromContinuations the "ok" handler on the queue is never called and voila you have an issue. Worse because technically there isn't even any blocking or threads waiting you just see an app that is "doing nothing" in debug with no reason why.

The same problem also affects the readLoop sadly in the same test. I have to fix both to make my app work.

Question: Why bother with starting a thread this way? What is the purposes of the queues? If it is just to make sure you aren't writing/reading simultaneously there's better ways using normal Async's without the WithContinuations method which is my experience is error-prone.

**Comment by ademar** [comment]
Hi, thanks for the report. Sure a PR is welcome :) 

You are correct the locks are not set up correctly . The purpose of the queues is to serialize reads and writes.

**Comment by mvkra** [comment]
If/when this is merged would appreciate a bug fix release (1.1.4) so that it can be used in place of the custom Suave logic we have at the moment. Thanks
